### ISSUE TITLE
Question About Error Handling

### ISSUE THREAD

**Issue by mlr-msft** [Marwes/combine#198]
Ethereum bytecode is represented with a string of hex digits-- e.g.:

`0x01020304...`

As such, there need to be an even number of hexadecimal digits in order for the parsing to succeed.

Through examining the examples, I've come up with the following code to parse the string:

```
use combine::{
    error::ParseError,
    many1,
    parser::char::{hex_digit, string},
    stream::Stream,
    Parser,
};

pub fn code<I>() -> impl Parser<Input = I, Output = Vec<u8>>
where
    I: Stream<Item = char>,
    // note: author of combine claims the following is necessary due
    // to rust-lang/rust#24159
    I::Error: ParseError<I::Item, I::Range, I::Position>,
{
    string("0x")
        .with(many1(hex_digit()).map(|s: String| hex::decode(s).unwrap()))
}
```

This works, but panics if an odd number of digits are provided. It's unclear to me how I would return a `ParseError<>` if the number of digits were odd, given that `map` does not accept a lambda that returns a `Result<>` type that would be checked by another combinator.

What's the correct way to deal with this situation without relying on panicking?

Thanks in advance.

**Comment by Lythenas** [comment]
I think something like `many1((hex_digit(), hex_digit()))` or `many1(count(2, hex_digit()))` instead of `many1(hex_digit())` could work. This ensures that you always take two hex digits instead of just one.

Another possibility is to use the `Parser::and_then` method instead of `map`. This method lets you return a `Result<O, E>`. `E` has to be convertible to the correct parse error.

**Comment by Marwes** [comment]
> I think something like many1((hex_digit(), hex_digit())) or many1(count(2, hex_digit())) instead of many1(hex_digit()) could work. This ensures that you always take two hex digits instead of just one.

Yep, this would be represented by `recognize(skip_many((hex_digit(), hex_digit())))`
